Mathlib/RingTheory/AlgebraicIndependent/TranscendenceBasis.lean

```diff
@@ -125,3 +208,299 @@ theorem IsTranscendenceBasis.isAlgebraic_field {F E : Type*} {x : ι → E}
   haveI : IsScalarTower (adjoin F S) (IntermediateField.adjoin F S) E :=
     IsScalarTower.of_algebraMap_eq (congrFun rfl)
   exact Algebra.IsAlgebraic.extendScalars (R := adjoin F S) (Subalgebra.inclusion_injective _)
+
+namespace AlgebraicIndependent
+
+variable (inj : Injective (algebraMap R A))
+
+section
+
+variable [NoZeroDivisors A]
+
+private def indepMatroid : IndepMatroid A where
+  E := univ
+  Indep := Set.AlgebraicIndependent R
+  indep_empty := (algebraicIndependent_empty_iff ..).mpr inj
+  indep_subset _ _ := (·.mono)
+  indep_aug I B I_ind h B_base := by
+    contrapose! h
+    rw [← isTranscendenceBasis_iff_maximal] at B_base ⊢
+    cases subsingleton_or_nontrivial R
+    · rw [isTranscendenceBasis_iff_of_subsingleton] at B_base ⊢
+      contrapose! h
+      have ⟨b, hb⟩ := B_base
+      exact ⟨b, ⟨hb, fun hbI ↦ h ⟨b, hbI⟩⟩, .of_subsingleton⟩
+    rw [I_ind.isTranscendenceBasis_iff_isAlgebraic]
+    replace B_base := B_base.isAlgebraic
+    rw [Subtype.range_val] at B_base ⊢
+    refine ⟨fun a ↦ (B_base.1 a).adjoin_of_forall_isAlgebraic fun x hx ↦ ?_⟩
+    contrapose! h
+    exact ⟨x, hx, I_ind.insert h⟩
+  indep_maximal X _ I ind hIX := exists_maximal_algebraicIndependent I X hIX ind
+  subset_ground _ _ := subset_univ _
+
+/-- If `R` is a commutative ring and `A` is a commutative `R`-algebra with injective algebra map
+and no zero-divisors, then the `R`-algebraic independent subsets of `A` form a matroid. -/
+def matroid : Matroid A := (indepMatroid inj).matroid.copyBase univ
+  (fun s ↦ IsTranscendenceBasis R ((↑) : s → A)) rfl
+  (fun B ↦ by simp_rw [Matroid.base_iff_maximal_indep, isTranscendenceBasis_iff_maximal]; rfl)
+
+instance : (matroid inj).Finitary where
+  indep_of_forall_finite := algebraicIndependent_of_finite
+
+@[simp] theorem matroid_e : (matroid inj).E = univ := rfl
+
+theorem matroid_indep_iff {s : Set A} :
+    (matroid inj).Indep s ↔ s.AlgebraicIndependent R := Iff.rfl
+
+theorem matroid_base_iff {s : Set A} :
+    (matroid inj).Base s ↔ IsTranscendenceBasis R ((↑) : s → A) := Iff.rfl
+
+theorem matroid_cRank_eq : (matroid inj).cRank = trdeg R A :=
+  (trdeg_eq_iSup_cardinalMk_isTranscendenceBasis _).symm
+
+end
+
+theorem matroid_basis_iff [IsDomain A] {s t : Set A} : (matroid inj).Basis s t ↔
+    s.AlgebraicIndependent R ∧ s ⊆ t ∧ ∀ a ∈ t, IsAlgebraic (adjoin R s) a := by
+  rw [Matroid.Basis, maximal_iff_forall_insert fun s t h hst ↦ ⟨h.1.subset hst, hst.trans h.2⟩]
+  simp_rw [matroid_indep_iff, ← and_assoc, matroid_e, subset_univ, and_true]
+  exact and_congr_right fun h ↦ ⟨fun max a ha ↦ of_not_not fun tr ↦ max _
+    (fun ha ↦ tr (isAlgebraic_algebraMap (⟨a, subset_adjoin ha⟩ : adjoin R s)))
+      ⟨.insert h.1 tr, insert_subset ha h.2⟩,
+    fun alg a ha h ↦ ((insert_iff ha).mp h.1).2 (alg _ <| h.2 <| mem_insert ..)⟩
+
+theorem matroid_basis_iff_of_subsingleton [Subsingleton A] {s t : Set A} :
```

Comment:
This lemma is an ad hoc consequence of the fact that your matroid is `Loopless`; i.e. singletons are independent.  (see https://github.com/apnelson1/Matroid/blob/main/Matroid/Loop.lean )

Reply:
Ah, I've got this wrong - it is a consequence of the fact that there are no nontrivial parallel pairs, even though there is a loop (zero) element.